File: smalivm/src/test/java/org/cf/smalivm/opcode/ThrowOpTest.java
//package org.cf.smalivm.opcode;
//
//import static org.junit.jupiter.api.Assertions.assertEquals;
//
//import org.cf.smalivm.VMState;
//import org.cf.smalivm.VMTester;
//import org.cf.smalivm.context.ExecutionGraph;
//import org.cf.smalivm.context.HeapItem;
//import org.cf.smalivm.context.MethodState;
//import org.cf.util.ClassNameUtils;
//import org.junit.jupiter.api.BeforeEach;
//import org.junit.jupiter.api.Test;
//
//public class ThrowOpTest {
//
//    private static final String CLASS_NAME = "Lthrow_test;";
//
//    private VMState expected;
//    private VMState initial;
//
//    @BeforeEach
//    public void setUp() {
//        expected = new VMState();
//        initial = new VMState();
//    }
//
//    @Test
//    public void canThrowNullPointerException() {
//        ExecutionGraph graph = VMTester.execute(CLASS_NAME, "throwNullPointerException()V", initial);
//
//        Class<?> exceptionClass = NullPointerException.class;
//        HeapItem item = graph.getTerminatingRegisterConsensus(0);
//        assertEquals(exceptionClass, item.getValue().getClass());
//        assertEquals(ClassNameUtils.toInternal(exceptionClass), item.getType());
//
//        HeapItem throwItem = graph.getTerminatingRegisterConsensus(MethodState.ThrowRegister);
//        assertEquals(item, throwItem);
//
//        VMTester.test(CLASS_NAME, "throwNullPointerException()V", expected);
//    }
//
//}

File: smalivm/src/test/java/org/cf/smalivm/opcode/ThrowOpTest.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.TestState
import org.cf.smalivm.Tester
import org.cf.smalivm.ExecutionState
import org.cf.util.ClassNameUtils
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class ThrowOpTest {
    private lateinit var expected: TestState
    private lateinit var initial: TestState

    companion object {
        private const val CLASS_NAME = "Lthrow_test;"
    }

    @BeforeEach
    fun setUp() {
        expected = TestState()
        initial = TestState()
    }

    @Test
    fun canThrowNullPointerException() {
        val graph = Tester.execute(CLASS_NAME, "throwNullPointerException()V", initial)
        val item = graph.getTerminatingRegisterConsensus(0)
        assertTrue(item!!.raw is NullPointerException)
        assertEquals(ClassNameUtils.toInternal(NullPointerException::class.java), item.type)
        val throwItem = graph.getTerminatingRegisterConsensus(ExecutionState.THROW_REGISTER)
        assertEquals(item, throwItem)
        Tester.test(CLASS_NAME, "throwNullPointerException()V", expected)
    }
}
